### ISSUE TITLE
Unable to build successfully with HIP backends

### ISSUE THREAD

**Issue by gsitaram** [cp2k/cp2k#2349]
When I try to include the flags `-D__OFFLOAD_HIP`, my compilation fails. The current build system is very confusing and the following are big concerns in my mind:
- Now we are including all HIP or CUDA backends by default and are having to exclude different backends if we choose to instead of following the general practice of choosing to include only the backends we want to have. This results in files not needed getting included in builds (see examples below).
- When I build for AMD GPUs, the build system seems to include the CUDA backend also? This could be a lack of understanding on my part, but this is causing problems.

Here is an example of building with `-D__NO_OFFLOAD_PW -D__OFFLOAD_HIP`. Note that `src/dbm/*.cu` and `src/grid/gpu/*.cu` files are being built by hipcc. I also see the `src/grid/hip/*` files being built with hipcc, but there is no problem there. I am able to exclude the dbm backend using `__NO_OFFLOAD_DBM` but the problem with the `grid/gpu` files remains.
```
[2022-10-12T00:08:26.459Z] hipcc -c -D__FFTW3 -D__parallel -D__COSMA -D__SCALAPACK -D__LIBXSMM -D__NO_OFFLOAD_PW -D__OFFLOAD_HIP -D__HIP_PLATFORM_AMD__ -I/opt/fftw/include  -D__LIBINT -D__MAX_CONTR=4 -D__LIBXC --amdgpu-target=gfx90a,gfx908,gfx906 -O3 -fdenormal-fp-math=ieee -fcuda-flush-denormals-to-zero -munsafe-fp-atomics -fPIC --std=c++11 /opt/cp2k/src/dbm/dbm_multiply_gpu.cu
[2022-10-12T00:08:26.962Z] hipcc -c -D__FFTW3 -D__parallel -D__COSMA -D__SCALAPACK -D__LIBXSMM -D__NO_OFFLOAD_PW -D__OFFLOAD_HIP -D__HIP_PLATFORM_AMD__ -I/opt/fftw/include  -D__LIBINT -D__MAX_CONTR=4 -D__LIBXC --amdgpu-target=gfx90a,gfx908,gfx906 -O3 -fdenormal-fp-math=ieee -fcuda-flush-denormals-to-zero -munsafe-fp-atomics -fPIC --std=c++11 /opt/cp2k/src/grid/gpu/grid_gpu_task_list.cu
[2022-10-12T00:08:27.213Z] [91mIn file included from /opt/cp2k/src/dbm/dbm_multiply_gpu.cu:15:
[2022-10-12T00:08:27.213Z] In file included from /opt/cp2k/src/dbm/dbm_multiply_gpu.h:15:
[2022-10-12T00:08:27.213Z] In file included from /opt/cp2k/src/dbm/dbm_multiply_internal.h:12:
[2022-10-12T00:08:27.213Z] /opt/cp2k/src/dbm/dbm_shard.h:10:10: fatal error: 'omp.h' file not found
[2022-10-12T00:08:27.213Z] #include <omp.h>
[2022-10-12T00:08:27.213Z]          ^~~~~~~
[2022-10-12T00:08:27.213Z] [0m[91m1 error generated when compiling for gfx906.
[2022-10-12T00:08:27.213Z] [0m[91mmake[3]: *** [/opt/cp2k/Makefile:521: dbm_multiply_gpu.o] Error 1
[2022-10-12T00:08:27.213Z] make[3]: *** Waiting for unfinished jobs....
[2022-10-12T00:08:27.465Z] [0m[91m/opt/cp2k/src/grid/gpu/grid_gpu_task_list.cu:12:10: fatal error: 'omp.h' file not found
[2022-10-12T00:08:27.465Z] #include <omp.h>
[2022-10-12T00:08:27.465Z]          ^~~~~~~
[2022-10-12T00:08:27.465Z] [0m[91m1 error generated when compiling for gfx906.
```
These `*.cu` files have some openmp pragmas for which we need to include the `-fopenmp` flag when compiling them. There is a problem with including `-fopenmp` in files compiled with `hipcc` because `hipcc` pulls in the openmp library found in llvm during linking. And since I use GNU compilers for everything else, we end up with `libomp.so` and `libgomp.so` in the final linked binary and performance drops.

If we can remove those openmp pragmas in these files or separate device code and host code into different translation units (compiled with different compilers), that would be one solution. I'd rather not build these files if they are not needed anyway.

**Comment by oschuett** [comment]
> Now we are including all HIP or CUDA backends by default ... This results in files not needed getting included in builds

Yes, but those unused files should get emptied by the pre-processor. However, in the case of `-D__NO_OFFLOAD_PW` there were indeed two `ifdef` statements missing.

> When I build for AMD GPUs, the build system seems to include the CUDA backend also?

No, but we're using `.cu` files for both CUDA and HIP. This allows us to have almost no code-duplication.

> There is a problem with including -fopenmp in files compiled with hipcc because hipcc pulls in the openmp library found in llvm during linking.

Oh, this is an unexpected deviation from CUDA where we can use GCC also as host compiler. Fortunately, it was not that difficult to move the OpenMP depended code to a different translation unit.

I've opened #2350 with those changes.

**Comment by mtaillefumier** [comment]
> > Now we are including all HIP or CUDA backends by default ... This results in files not needed getting included in builds
> 
> Yes, but those unused files should get emptied by the pre-processor. However, in the case of `-D__NO_OFFLOAD_PW` there were indeed two `ifdef` statements missing.

If these files are not needed for a given compilation, it is best to exclude  them totally from the compilation. 
- the -D__NO_OFFLOAD_PW (GRID, DBM, etc) *does* not follow the cp2k convention for all the other optional packages (-D__ELPA, etc...). 
- The -D__OFFLOAD_PW (etc) makes far more sense 

> > When I build for AMD GPUs, the build system seems to include the CUDA backend also?
> 
> No, but we're using `.cu` files for both CUDA and HIP. This allows us to have almost no code-duplication.

The question was inferring the following point. The nvidia backend for the grid module is included  in cp2k build supporting AMD hardware only. see this line here **https://github.com/cp2k/cp2k/blob/cea20fb160686b319917257b3a6652216704cc48/src/grid/grid_task_list.c#L95** and below. 

including any code targeting NVIDIA hardware to a CP2K build targeting AMD hardware creates additional confusion

**Comment by oschuett** [comment]
> If these files are not needed for a given compilation, it is best to exclude them totally from the compilation.

Yes, but it would add complexity to the already tricky build process. A few mostly empty translation units don't cause any real problems. I assume this will become easier once we've switched to cmake?

> The -D__OFFLOAD_PW (etc) makes far more sense

As we added more GPU kernels it became unreasonable to ask the user to enable them all individually. So, instead we're now treating GPU acceleration like MPI - it's either enabled or disabled. Those `__NO_OFFLOAD_foo` flags are really only meant for debugging.

> The nvidia backend for the grid module is included in cp2k build supporting AMD hardware only.

Yes, that's on purpose because the GPU backend should work on AMD hardware as well. Its performance is probably poor, but it should give correct results. (Since I don't have access to the hardware I don't actually know.)

In the coming months I plan to refactor the GPU backend to address #1785. In the process it'll become more similar to your HIP backend. My hope is that we can eventually merge the two and serve both vendors with a single backend.

**Comment by gsitaram** [comment]
I also agree with the sentiment that only the files intended to be used should be included for compilation. In cases where the files differ based on the hardware used, why include the files that are not going to be traversed at all? This causes confusion. I am looking forward to the CMake build system as well so such problems can be avoided.

Let me give an example. I tried to turn off all GPU backends by not setting `-D__OFFLOAD_HIP` and got these errors:
```
hipcc -c -D__FFTW3 -D__parallel -D__COSMA -D__SCALAPACK -D__LIBXSMM  -I/opt/fftw/include  -D__LIBINT -D__MAX_CONTR=4 -D__LIBXC --amdgpu-target=gfx90a,gfx908,gfx906 -O3 -fdenormal-fp-math=ieee -fcuda-flush-denormals-to-zero -munsafe-fp-atomics -fPIC --std=c++11 /opt/cp2k/src/pw/gpu/pw_gpu_kernels.cu
In file included from /opt/cp2k/src/pw/gpu/pw_gpu_kernels.cu:8:
/opt/cp2k/src/pw/gpu/pw_gpu_kernels.h:22:53: error: unknown type name 'offloadStream_t'
                                   const int ngpts, offloadStream_t stream);
/opt/cp2k/src/pw/gpu/pw_gpu_kernels.cu:22:20: error: use of undeclared identifier 'blockIdx'
  const int igpt = blockIdx.x * blockDim.x + threadIdx.x;
                   ^
```
This should not happen. GPU code must not be included if we want to compile for the CPU. 

After debugging, I figured out that if OFFLOAD_CC is set (totally not expected to be used in this scenario), GPU sources are included.

**Comment by oschuett** [comment]
> This should not happen. GPU code must not be included if we want to compile for the CPU.

Yes, this was indeed a bug. I fixed it earlier today via https://github.com/cp2k/cp2k/commit/655fccdd2bc6c9204520ab025ebd8d2beb06e4f9.

> This causes confusion.

Look, I'm not saying our build system is perfect. It's simply the only one we have at the moment. It was conceived 20+ years ago when things were simpler....

> I am looking forward to the CMake build system as well so such problems can be avoided.

Yeah, me too. In fact, I'm holding off on further investments into our current build system in anticipation of cmake.

**Comment by mtaillefumier** [comment]
> > If these files are not needed for a given compilation, it is best to exclude them totally from the compilation.
> 
> Yes, but it would add complexity to the already tricky build process. A few mostly empty translation units don't cause any real problems. I assume this will become easier once we've switched to cmake?

I solved this particular issue by only including the files related to a given option. For instance -DUSE_OPTION1 would trigger the complication of any translation unit associated to it. it is done by adding the translation files to the list of files to be compiled. The only tedious part is to add the file to the CMakeLists.txt but it is done only one time.

The cmake build system is working. We use it internally for almost all our tests (for lumi, preliminary work on dlaf, etc). It is ready to be tested more extensively if users are interested in. As always with all these things, problems come from corners cases, for instance cosma compiled a multi threaded blas while cp2k wants a single threaded version (the default choice). It is not an issue with mkl but can be problematic with libsci or openblas.

So far (excluding elpa dependency for now, the naming convention makes things more complicated), cp2k can be build with cmake on the following systems (with or without cuda, hip depends on rocm version but still possible)
- Cray environments with mkl or libsci (tested with gcc, but there is almost no difference with the others programming envs)
- Summit when we explicitly indicate what the blas libraries are
- Fedora (redhat) with mkl, openblas, flexiblas (need more testing for the last one). Should work as well in ubuntu as almost everything is detected automatically (the main problem will always be blas and lapack).

Using cmake was also beneficial for dbcsr as I found few inconsistencies in its build system. 
 
> > The -D__OFFLOAD_PW (etc) makes far more sense
> 
> As we added more GPU kernels it became unreasonable to ask the user to enable them all individually. So, instead we're now treating GPU acceleration like MPI - it's either enabled or disabled. Those `__NO_OFFLOAD_foo` flags are really only meant for debugging.

These options will only be visible to the user if they explicitly say -DCP2K_DEV_OPTIONS=ON. it will remove a lot of issues.

**Comment by oschuett** [comment]
> The cmake build system is working. 

That's great news!

>  As always with all these things, problems come from corners cases,

Yes, it's the unfortunate nature of build systems that they have to deal with the combinatorial explosion created by build options and build environments. Therefore this will be one of those slow migrations where both systems have to coexist for a while. I'd also like to make at least one release in between that includes both build systems. All in all it will probably take the better part of a year before we can eventually remove our old Makefile. Nevertheless, I'm eager to get started. So, feel free to open a PR when you're ready.

**Comment by mtaillefumier** [comment]
> 

then I will most probably make a PR this week as it is a bit tedious to keep this out of the main branch. I also agree the two systems need to coexist for one or two versions. I did not modify anything in the code and used all the options already present to turn on and off dependencies.

**Comment by gsitaram** [comment]
@oschuett , thanks for the quick fix. I can confirm that my build is successful now using your updated code.
